universal/include/userver/formats/universal/universal.hpp

```diff
@@ -0,0 +1,235 @@
+    return (args + ...);
+  } else {
+    return {};
+  }
+}
+
+
+
+template <typename T, typename FieldConfigType>
+inline constexpr std::string UniversalCheckField(
+          T&& response,
+          FieldConfig<FieldConfigType> config) {
+  using FieldType = std::remove_cvref_t<decltype(response)>;
+  auto error = [&]<auto... Is>(std::index_sequence<Is...>) -> std::string {
+    auto f = [&]<auto II>(Wrapper<II>) -> std::string {
+      auto check = boost::pfr::get<II>(config);
+      if constexpr(HasCheck<decltype(RemoveOptional(check)), decltype(RemoveOptional(response))>::value) {
+        if constexpr(meta::kIsOptional<FieldType>) {
+          if(response && check && !check->Check(RemoveOptional(response)).empty()) {
+            return check->Check(RemoveOptional(response));
+          }
+        } else {
+          if(check && !check->Check(response).empty()) {
+            return check->Check(response);
+          }
+        }
+      }
+      return "";
+    };
+    return add<std::string>(f(Wrapper<Is>{})...);
+  }(std::make_index_sequence<boost::pfr::tuple_size_v<decltype(config)>>());
+  if constexpr(meta::kIsOptional<std::remove_cvref_t<T>>) {
+    if(response) {
+      error += config.Check(*response);
+    }
+  } else {
+    error += config.Check(std::forward<T>(response));
+  }
+  return error;
+}
+
+
+template <typename MainClass, auto I, typename T>
+inline constexpr auto UniversalParseField(
+          T&& value,
+          FieldConfig<kFieldTypeOnIndex<MainClass, I>> config) {
+  auto response = config.template Read<MainClass, I>(value);
+  const auto error = UniversalCheckField(response, config);
+  if(!error.empty()) {
+    throw std::runtime_error(error);
+  }
+  return response;
+}
+
+template <typename MainClass, auto I, typename T>
+inline constexpr auto UniversalTryParseField(
+          T&& value,
+          FieldConfig<kFieldTypeOnIndex<MainClass, I>> config) {
+  auto response = config.template TryRead<MainClass, I>(value);
+  return [&]() -> decltype(response) {
+    if(UniversalCheckField(response, std::move(config)).empty()) {
+      return response;
+    }
+    return std::nullopt;
+  }();
+}
+
+template <typename MainClass, auto I, typename T, typename Builder>
+inline constexpr auto UniversalSerializeField(
+        T&& value,
+        Builder& builder,
+        FieldConfig<kFieldTypeOnIndex<MainClass, I>> config) {
+  UniversalCheckField(value, config);
+  config.template Write(std::forward<T>(value), boost::pfr::get_name<I, MainClass>(), boost::pfr::names_as_array<MainClass>(), builder);
+}
+struct EmptyCheck {
+  template <typename Value>
+  inline constexpr std::string Check(Value&&) const {
+    return "";
+  }
+};
+} // namespace impl
+
+template <typename T>
+impl::Disabled kSerialization;
+
+template <typename T>
+auto kDeserialization = kSerialization<T>;
+
+
+template <typename T>
+struct SerializationConfig {
+  using kFieldsConfigType = decltype([]<auto... I>(std::index_sequence<I...>){
```

Comment:
This is actually `public` due to `struct`. Should probably make it `class` since the field is `private`